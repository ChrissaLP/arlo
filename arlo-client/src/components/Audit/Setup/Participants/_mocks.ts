const jurisdictionFile = new File(
  [
    'Jurisdiction","Admin Email"',
    '"Death Star","wtarkin@empire.gov"',
    '"Hoth","admin@rebelalliance.ninja"',
    '"Tatooine","jabba@hutt.biz"',
  ],
  'jurisdictions.csv',
  { type: 'text/csv' }
)

export default jurisdictionFile
