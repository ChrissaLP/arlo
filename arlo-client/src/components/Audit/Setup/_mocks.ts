import { ElementType } from '../../../types'
import { setupStages } from './index'
import { ISidebarMenuItem } from '../../Atoms/Sidebar'

const relativeStages = (
  stage: ElementType<typeof setupStages>,
  state: ISidebarMenuItem['state'] = 'live'
) => {
  const prevTitle = setupStages[setupStages.indexOf(stage) - 1]
  const prevStage: ISidebarMenuItem = {
    title: prevTitle,
    active: false,
    activate: jest.fn(),
    state,
  }
  const nextTitle = setupStages[setupStages.indexOf(stage) + 1]
  const nextStage: ISidebarMenuItem = {
    title: nextTitle,
    active: false,
    activate: jest.fn(),
    state,
  }
  return { prevStage, nextStage }
}

export default relativeStages
