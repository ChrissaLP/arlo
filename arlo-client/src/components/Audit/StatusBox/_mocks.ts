import {
  IJurisdictionsResponse,
  FileProcessingStatus,
} from './getJurisdictions'
import { IRound } from '../../../types'

export const jurisdictionMocks: {
  [key in
    | 'none'
    | 'noUploads'
    | 'halfUploads'
    | 'allFinishedFirstRound']: IJurisdictionsResponse
} = {
  none: { jurisdictions: [] },
  noUploads: {
    jurisdictions: [
      {
        id: 'jurisdiction-1',
        name: 'Jurisdiction One',
        currentRoundStatus: null,
        ballotManifest: {
          file: null,
          numBallots: null,
          numBatches: null,
          processing: null,
        },
      },
      {
        id: 'jurisdiction-2',
        name: 'Jurisdiction Two',
        currentRoundStatus: null,
        ballotManifest: {
          file: null,
          numBallots: null,
          numBatches: null,
          processing: null,
        },
      },
    ],
  },
  halfUploads: {
    jurisdictions: [
      {
        id: 'jurisdiction-1',
        name: 'Jurisdiction One',
        currentRoundStatus: null,
        ballotManifest: {
          file: 'ballot manifest',
          numBallots: 10,
          numBatches: 1,
          processing: FileProcessingStatus.Processed,
        },
      },
      {
        id: 'jurisdiction-2',
        name: 'Jurisdiction Two',
        currentRoundStatus: null,
        ballotManifest: {
          file: null,
          numBallots: null,
          numBatches: null,
          processing: null,
        },
      },
    ],
  },
  allFinishedFirstRound: {
    jurisdictions: [
      {
        id: 'jurisdiction-1',
        name: 'Jurisdiction One',
        currentRoundStatus: 1,
        ballotManifest: {
          file: 'ballot manifest',
          numBallots: 10,
          numBatches: 1,
          processing: FileProcessingStatus.Processed,
        },
      },
      {
        id: 'jurisdiction-2',
        name: 'Jurisdiction Two',
        currentRoundStatus: 1,
        ballotManifest: {
          file: 'ballot manifest',
          numBallots: 10,
          numBatches: 1,
          processing: FileProcessingStatus.Processed,
        },
      },
    ],
  },
}

export const roundMocks: {
  [key in
    | 'none'
    | 'firstRoundComplete'
    | 'firstRoundIncomplete'
    | 'secondRoundIncomplete']: { rounds: IRound[] }
} = {
  none: { rounds: [] },
  firstRoundIncomplete: {
    rounds: [
      {
        contests: [
          {
            endMeasurements: {
              isComplete: null,
              pvalue: null,
            },
            id: 'contest-1',
            results: {},
            sampleSize: null,
            sampleSizeOptions: null,
          },
        ],
        endedAt: null,
        startedAt: '2019-07-18T16:34:07.000Z',
        id: 'round-1',
      },
    ],
  },
  firstRoundComplete: {
    rounds: [
      {
        contests: [
          {
            endMeasurements: {
              isComplete: true,
              pvalue: 0.00020431431380638307,
            },
            id: 'contest-1',
            results: {
              'choice-1': 100,
              'choice-2': 167,
            },
            sampleSize: 379,
            sampleSizeOptions: [
              { size: 269, type: 'ASN', prob: null },
              { size: 379, prob: 0.8, type: null },
              { size: 78, prob: null, type: null },
            ],
          },
        ],
        endedAt: '2019-07-18T16:59:34.000Z',
        startedAt: '2019-07-18T16:34:07.000Z',
        id: 'round-1',
      },
    ],
  },
  secondRoundIncomplete: {
    rounds: [
      {
        contests: [
          {
            endMeasurements: {
              isComplete: false,
              pvalue: 1,
            },
            id: 'contest-1',
            results: {
              'choice-1': 0,
              'choice-2': 0,
            },
            sampleSize: null,
            sampleSizeOptions: [
              { size: 269, type: 'ASN', prob: null },
              { size: 379, prob: 0.8, type: null },
              { size: 78, prob: null, type: null },
            ],
          },
        ],
        endedAt: '2019-07-18T16:35:07.000Z',
        startedAt: '2019-07-18T16:34:07.000Z',
        id: 'round-1',
      },
      {
        contests: [
          {
            endMeasurements: {
              isComplete: null,
              pvalue: null,
            },
            id: 'contest-1',
            results: {},
            sampleSize: null,
            sampleSizeOptions: [
              { size: 269, type: 'ASN', prob: null },
              { size: 379, prob: 0.8, type: null },
              { size: 78, prob: null, type: null },
            ],
          },
        ],
        endedAt: null,
        startedAt: '2019-07-18T16:34:07.000Z',
        id: 'round-2',
      },
    ],
  },
}
