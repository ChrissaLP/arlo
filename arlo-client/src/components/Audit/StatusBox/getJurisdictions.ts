import { toast } from 'react-toastify'
import { api, checkAndToast } from '../../utilities'
import { IErrorResponse } from '../../../types'

export enum FileProcessingStatus {
  Blank = 'NULL', // only returned from getJurisdictionFileStatus, represents the null state from the server
  ReadyToProcess = 'READY_TO_PROCESS', // only received from the server, never returned from getJurisdictionFileStatus; equivalent to 'Processing'
  Processing = 'PROCESSING',
  Processed = 'PROCESSED',
  Errored = 'ERRORED',
}

export interface IJurisdictionsResponse {
  jurisdictions: {
    id: string
    name: string
    ballotManifest: {
      file: string | null
      processing: FileProcessingStatus | null
      numBallots: number | null
      numBatches: number | null
    }
    currentRoundStatus: number | null
  }[]
}

const getJurisdictions = async (
  electionId: string
): Promise<undefined | IJurisdictionsResponse> => {
  try {
    const jurisdictionsOrError:
      | IJurisdictionsResponse
      | IErrorResponse = await api(`/election/${electionId}/jurisdiction`)
    // checkAndToast left here for consistency and reference but not tested since it's vestigial
    /* istanbul ignore next */
    if (checkAndToast(jurisdictionsOrError)) {
      return undefined
    }
    return jurisdictionsOrError
  } catch (err) /* istanbul ignore next */ {
    // TEST TODO
    toast.error(err.message)
    return undefined
  }
}

export default getJurisdictions
