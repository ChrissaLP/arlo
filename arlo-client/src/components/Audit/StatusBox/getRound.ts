import { toast } from 'react-toastify'
import { api, checkAndToast } from '../../utilities'
import { IRound, IErrorResponse } from '../../../types'

const getRound = async (electionId: string): Promise<IRound[]> => {
  try {
    const roundsOrError: { rounds: IRound[] } | IErrorResponse = await api(
      `/election/${electionId}/round`
    )
    // checkAndToast left here for consistency and reference but not tested since it's vestigial
    /* istanbul ignore next */
    if (checkAndToast(roundsOrError) || !roundsOrError.rounds.length) {
      return []
    }
    return roundsOrError.rounds
  } catch (err) /* istanbul ignore next */ {
    // TEST TODO
    toast.error(err.message)
    return []
  }
}

export default getRound
