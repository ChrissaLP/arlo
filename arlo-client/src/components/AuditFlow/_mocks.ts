import { IAuditBoard } from '../../types'

export const dummyBoard: IAuditBoard[] = [
  {
    id: '123',
    name: 'Audit Board #1',
    members: [],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
    ballots: [
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
        id: '1',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
        id: '2',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
        id: '3',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'NOT_AUDITED',
        vote: null,
        comment: '',
        id: '4',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '5',
      },
    ],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
    ballots: [
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '1',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '2',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '3',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '4',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '5',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '6',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '7',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '8',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '9',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '10',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '11',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '12',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '13',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '14',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '15',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '16',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '17',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '18',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '19',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '20',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '21',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '22',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '23',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '24',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '25',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '26',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '27',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '28',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '29',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '30',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '31',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '32',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '33',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '34',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '35',
      },
      {
        tabulator: '11',
        batch: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        position: '313',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '36',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '37',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '38',
      },
      {
        tabulator: '17',
        batch: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        position: '2112',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '39',
      },
      {
        tabulator: '23',
        batch: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        position: '1789',
        status: 'AUDITED',
        vote: null,
        comment: '',
        id: '40',
      },
    ],
  },
]

export default {
  dummyBoard,
}
