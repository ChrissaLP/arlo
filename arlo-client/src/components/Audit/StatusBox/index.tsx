import React, { useEffect, useState } from 'react'
import styled from 'styled-components'
import { Callout, H4 } from '@blueprintjs/core'
import FormButton from '../../Atoms/Form/FormButton'
import getRound from './getRound'
import getJurisdictions, { IJurisdictionsResponse } from './getJurisdictions'

const Wrapper = styled(Callout)`
  .details {
    width: 50%;
  }
  .wrapper {
    margin: 10px 0 10px 0;
    text-align: right;
  }
`

const generateStatuses = (
  totalJurisdictions: number,
  manifestsUploaded: number,
  roundsCompleted: number,
  round: number,
  launched: boolean,
  complete: boolean
): [
  string,
  string[],
  'NO_BUTTON' | 'START_ROUND_BUTTON' | 'DOWNLOAD_BUTTON'
] => {
  if (complete) {
    // all jurisdictions have completed the audit
    return ['The audit is complete.', [], 'DOWNLOAD_BUTTON']
  }
  if (!complete && roundsCompleted === totalJurisdictions) {
    // jurisdictions have finished the round, but another is needed
    return [
      `Round ${round} of the audit is complete - another round needed.`,
      [`When you are ready, start Round ${round + 1}`],
      'START_ROUND_BUTTON',
    ]
  }
  if (launched) {
    // rounds have started
    return [
      `Round ${round} of the audit is in progress.`,
      [
        `${roundsCompleted} of ${totalJurisdictions} have completed Round ${round}`,
      ],
      'NO_BUTTON',
    ]
  }
  // the audit hasn't been launched yet
  return [
    'The audit has not started.',
    [
      'Audit setup is not complete.',
      `${manifestsUploaded} of ${totalJurisdictions} have completed file uploads.`,
    ],
    'NO_BUTTON',
  ]
}

interface IProps {
  electionId: string
}

const StatusBox = ({ electionId }: IProps) => {
  const downloadAuditReport = async (e: React.FormEvent) => {
    e.preventDefault()
    window.open(`/election/${electionId}/audit/report`)
  }

  const [currentRound, setCurrentRound] = useState(0)
  const [{ jurisdictions }, setJurisdictions] = useState<
    IJurisdictionsResponse
  >({ jurisdictions: [] })
  useEffect(() => {
    // get pertinent data
    ;(async () => {
      const round = await getRound(electionId)
      const jurisdictionList = await getJurisdictions(electionId)
      setCurrentRound(round)
      if (jurisdictionList) setJurisdictions(jurisdictionList)
    })()
  }, [electionId, setCurrentRound, setJurisdictions])
  console.log(jurisdictions)

  const [title, details, button] = generateStatuses(
    jurisdictions.length,
    jurisdictions.filter(j => j.ballotManifest.file).length,
    jurisdictions.filter(j => j.currentRoundStatus === currentRound).length,
    currentRound,
    currentRound > 0,
    false
  )

  return (
    <Wrapper>
      <H4>{title}</H4>
      <div className="details">
        {details.map(v => (
          <p key={v}>{v}</p>
        ))}
      </div>
      <div className="wrapper">
        {(() => {
          switch (button) {
            case 'DOWNLOAD_BUTTON':
              return (
                <FormButton intent="success" onClick={downloadAuditReport}>
                  Download Audit Reports
                </FormButton>
              )
            case 'START_ROUND_BUTTON': // button to submit to the rounds endpoint with a new round number
            case 'NO_BUTTON':
            default:
              return null
          }
        })()}
      </div>
    </Wrapper>
  )
}

export default StatusBox
