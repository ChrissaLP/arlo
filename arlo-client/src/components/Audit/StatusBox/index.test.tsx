import React from 'react'
import { render, wait, fireEvent } from '@testing-library/react'
import * as utilities from '../../utilities'
import StatusBox from '.'
import { generateApiMock } from '../../testUtilities'
import { jurisdictionMocks, roundMocks } from './_mocks'

const apiMock: jest.SpyInstance<
  ReturnType<typeof utilities.api>,
  Parameters<typeof utilities.api>
> = jest.spyOn(utilities, 'api').mockImplementation()

const checkAndToastMock: jest.SpyInstance<
  ReturnType<typeof utilities.checkAndToast>,
  Parameters<typeof utilities.checkAndToast>
> = jest.spyOn(utilities, 'checkAndToast').mockReturnValue(false)

checkAndToastMock.mockReturnValue(false)

afterEach(() => {
  apiMock.mockClear()
  checkAndToastMock.mockClear()
})

describe('StatusBox', () => {
  it('renders initial message', () => {
    apiMock.mockImplementation(
      generateApiMock({
        roundReturn: roundMocks.none,
        jurisdictionReturn: jurisdictionMocks.none,
      })
    )
    const { container, queryByText } = render(
      <StatusBox electionId="1" refreshId="1" />
    )
    expect(queryByText('The audit has not started.')).toBeTruthy()
    expect(queryByText('Audit setup is not complete.')).toBeTruthy()
    expect(queryByText('No jurisdictions have been created yet.')).toBeTruthy()
    expect(queryByText('Download Audit Reports')).toBeFalsy()
    expect(queryByText('Start Round 1')).toBeFalsy()
    expect(container).toMatchSnapshot()
  })

  it('after jurisdictions are made with no uploads', async () => {
    apiMock.mockImplementation(
      generateApiMock({
        roundReturn: roundMocks.none,
        jurisdictionReturn: jurisdictionMocks.noUploads,
      })
    )
    const { container, queryByText } = render(
      <StatusBox electionId="1" refreshId="1" />
    )
    await wait(() => {
      expect(queryByText('The audit has not started.')).toBeTruthy()
      expect(queryByText('Audit setup is not complete.')).toBeTruthy()
      expect(queryByText('0 of 2 have completed file uploads.')).toBeTruthy()
      expect(queryByText('Download Audit Reports')).toBeFalsy()
      expect(queryByText('Start Round 1')).toBeFalsy()
      expect(container).toMatchSnapshot()
    })
  })

  it('after jurisdictions are made with partial uploads', async () => {
    apiMock.mockImplementation(
      generateApiMock({
        roundReturn: { rounds: [] },
        jurisdictionReturn: jurisdictionMocks.halfUploads,
      })
    )
    const { container, queryByText } = render(
      <StatusBox electionId="1" refreshId="1" />
    )
    await wait(() => {
      expect(queryByText('The audit has not started.')).toBeTruthy()
      expect(queryByText('Audit setup is not complete.')).toBeTruthy()
      expect(queryByText('1 of 2 have completed file uploads.')).toBeTruthy()
      expect(queryByText('Download Audit Reports')).toBeFalsy()
      expect(queryByText('Start Round 1')).toBeFalsy()
      expect(container).toMatchSnapshot()
    })
  })

  it('after round one is started', async () => {
    apiMock.mockImplementation(
      generateApiMock({
        roundReturn: roundMocks.firstRoundIncomplete,
        jurisdictionReturn: jurisdictionMocks.halfUploads,
      })
    )
    const { container, queryByText } = render(
      <StatusBox electionId="1" refreshId="1" />
    )
    await wait(() => {
      expect(queryByText('Round 1 of the audit is in progress.')).toBeTruthy()
      expect(queryByText('0 of 2 have completed Round 1')).toBeTruthy()
      expect(queryByText('Download Audit Reports')).toBeFalsy()
      expect(queryByText('Start Round 1')).toBeFalsy()
      expect(container).toMatchSnapshot()
    })
  })

  it('after round one is finished but another is needed', async () => {
    apiMock.mockImplementation(
      generateApiMock({
        roundReturn: roundMocks.firstRoundIncomplete,
        jurisdictionReturn: jurisdictionMocks.allFinishedFirstRound,
      })
    )
    const { container, queryByText } = render(
      <StatusBox electionId="1" refreshId="1" />
    )
    await wait(() => {
      expect(
        queryByText('Round 1 of the audit is complete - another round needed.')
      ).toBeTruthy()
      expect(queryByText('When you are ready, start Round 2')).toBeTruthy()
      expect(queryByText('Download Audit Reports')).toBeFalsy()
      expect(queryByText('Start Round 2')).toBeTruthy()
      expect(container).toMatchSnapshot()
    })
  })

  it('after completion of audit in first round', async () => {
    apiMock.mockImplementation(
      generateApiMock({
        roundReturn: roundMocks.firstRoundComplete,
        jurisdictionReturn: jurisdictionMocks.halfUploads,
      })
    )
    const { container, queryByText } = render(
      <StatusBox electionId="1" refreshId="1" />
    )
    await wait(() => {
      expect(queryByText('The audit is complete.')).toBeTruthy()
      expect(queryByText('Download Audit Reports')).toBeTruthy()
      expect(queryByText('Start Round 1')).toBeFalsy()
      expect(container).toMatchSnapshot()
    })
  })

  it('downloads audit report', async () => {
    window.open = jest.fn()
    apiMock.mockImplementation(
      generateApiMock({
        roundReturn: roundMocks.firstRoundComplete,
        jurisdictionReturn: jurisdictionMocks.halfUploads,
      })
    )
    const { getByText, queryByText } = render(
      <StatusBox electionId="1" refreshId="1" />
    )
    await wait(() => expect(queryByText('Download Audit Reports')).toBeTruthy())
    fireEvent.click(getByText('Download Audit Reports'), { bubbles: true })
    await wait(() => {
      expect(window.open).toBeCalledTimes(1)
      expect(window.open).toBeCalledWith('/election/1/audit/report')
    })
  })

  it('starts round two', async () => {
    apiMock
      .mockImplementationOnce(
        generateApiMock({
          roundReturn: roundMocks.firstRoundIncomplete,
          jurisdictionReturn: jurisdictionMocks.allFinishedFirstRound,
        })
      )
      .mockImplementationOnce(
        generateApiMock({
          roundReturn: roundMocks.firstRoundIncomplete,
          jurisdictionReturn: jurisdictionMocks.allFinishedFirstRound,
        })
      )
      .mockImplementation(generateApiMock({ roundReturn: { status: 'ok' } }))
    const { getByText, queryByText } = render(
      <StatusBox electionId="1" refreshId="1" />
    )
    await wait(() => {
      expect(
        queryByText('Round 1 of the audit is complete - another round needed.')
      ).toBeTruthy()
      expect(queryByText('When you are ready, start Round 2')).toBeTruthy()
      expect(queryByText('Download Audit Reports')).toBeFalsy()
      expect(queryByText('Start Round 2')).toBeTruthy()
    })
    fireEvent.click(getByText('Start Round 2'), { bubbles: true })
    await wait(() => {
      expect(apiMock).toHaveBeenCalledTimes(3)
      expect(apiMock.mock.calls[2][0]).toBe('/election/1/round')
      expect(apiMock.mock.calls[2][1]).toMatchObject({
        body: JSON.stringify({
          roundNum: 2,
        }),
        headers: {
          'Content-Type': 'application/json',
        },
        method: 'POST',
      })
    })
  })
})
