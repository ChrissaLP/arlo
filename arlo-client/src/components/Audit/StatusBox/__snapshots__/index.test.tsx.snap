// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusBox after completion of audit in first round 1`] = `
<div>
  <div
    class="bp3-callout sc-htpNat hJnrNk"
  >
    <h4
      class="bp3-heading"
    >
      The audit is complete.
    </h4>
    <div
      class="details"
    />
    <div
      class="wrapper"
    >
      <button
        class="bp3-button bp3-intent-success sc-bwzfXH eBfcXh"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Download Audit Reports
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`StatusBox after jurisdictions are made with no uploads 1`] = `
<div>
  <div
    class="bp3-callout sc-htpNat hJnrNk"
  >
    <h4
      class="bp3-heading"
    >
      The audit has not started.
    </h4>
    <div
      class="details"
    >
      <p>
        Audit setup is not complete.
      </p>
      <p>
        0 of 2 have completed file uploads.
      </p>
    </div>
    <div
      class="wrapper"
    />
  </div>
</div>
`;

exports[`StatusBox after jurisdictions are made with partial uploads 1`] = `
<div>
  <div
    class="bp3-callout sc-htpNat hJnrNk"
  >
    <h4
      class="bp3-heading"
    >
      The audit has not started.
    </h4>
    <div
      class="details"
    >
      <p>
        Audit setup is not complete.
      </p>
      <p>
        1 of 2 have completed file uploads.
      </p>
    </div>
    <div
      class="wrapper"
    />
  </div>
</div>
`;

exports[`StatusBox after round one is finished but another is needed 1`] = `
<div>
  <div
    class="bp3-callout sc-htpNat hJnrNk"
  >
    <h4
      class="bp3-heading"
    >
      Round 1 of the audit is complete - another round needed.
    </h4>
    <div
      class="details"
    >
      <p>
        When you are ready, start Round 2
      </p>
    </div>
    <div
      class="wrapper"
    >
      <button
        class="bp3-button bp3-intent-success sc-bwzfXH eBfcXh"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Start Round 
          2
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`StatusBox after round one is started 1`] = `
<div>
  <div
    class="bp3-callout sc-htpNat hJnrNk"
  >
    <h4
      class="bp3-heading"
    >
      Round 1 of the audit is in progress.
    </h4>
    <div
      class="details"
    >
      <p>
        0 of 2 have completed Round 1
      </p>
    </div>
    <div
      class="wrapper"
    />
  </div>
</div>
`;

exports[`StatusBox renders initial message 1`] = `
<div>
  <div
    class="bp3-callout sc-htpNat hJnrNk"
  >
    <h4
      class="bp3-heading"
    >
      The audit has not started.
    </h4>
    <div
      class="details"
    >
      <p>
        Audit setup is not complete.
      </p>
      <p>
        No jurisdictions have been created yet.
      </p>
    </div>
    <div
      class="wrapper"
    />
  </div>
</div>
`;
